odl/discr/lp_discr.py

```diff
@@ -670,28 +837,35 @@ def conj(self, out=None):
             self.tensor.conj(out=out.tensor)
             return out
 
-    def __setitem__(self, indices, values):
-        """Set values of this element.
+    def __getitem__(self, indices):
+        """Return ``self[indices]``.
 
         Parameters
         ----------
-        indices : int or `slice`
-            The position(s) that should be set
-        values : scalar or `array-like`
-            Value(s) to be assigned.
-            If ``indices`` is an integer, ``values`` must be a scalar
-            value.
-            If ``indices`` is a slice, ``values`` must be
-            broadcastable to the size of the slice (same size,
-            shape ``(1,)`` or scalar).
-            For ``indices == slice(None)``, i.e. in the call
-            ``vec[:] = values``, a multi-dimensional array of correct
-            shape is allowed as ``values``.
+        indices : index expression
+            The position(s) that should be accessed.
+
+        Returns
+        -------
+        values : `DiscreteLpElement`
+            The value(s) at the index (indices).
+
+        Examples
+        --------
         """
-        if values in self.space:
-            self.tensor[indices] = values.tensor
+        if isinstance(indices, type(self)):
+            indices = indices.tensor.data
+        res_tens = self.tensor.__getitem__(indices)
+
+        try:
+            res_space = self.space[indices]
+        except (ValueError, TypeError):
+            # TODO: adapt error type
```

Comment:
We should be quite careful here, i guess this is OK for now, but we need to document it (similar to numpy arrays returning flat arrays in some cases, which is confusing). We also expect stuff like

```python
vec[complicated_indices] = 2 * vec[complicated_indices]
```

to work.

Reply:
That code would call `__setitem__` anyway and not be affected by this. But yeah, catching to many exceptions is a bit dangerous.

Reply:
I *think* there are unit tests with "complicated" indices for `__setitem__`, but I'll check again.